### PR TITLE
[Review] Support for Quantiles

### PR DESCRIPTION
**Pull request by razajafri**
These are the java bindings for calculating exact and approx quantiles

### CODE REVIEW

**Comment by razajafri** [comment]
I ran clang-format on the files which created some format changes PTAL

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 1]
```diff
@@ -1844,6 +1848,23 @@ public static ColumnVector concatenate(ColumnVector... columns) {
     return new ColumnVector(concatenate(columnHandles));
   }
 
+  /**
+   * Calculate the quantile of this ColumnVector
+   * @param method   the method used to calculate the quantile
+   * @param quantile the quantile value [0,1]
+   */
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java]: Can we add a `@return` statement to explain what is returned by the quantile?  Is it always a double?  Does it match the input type? etc.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 2]
```diff
@@ -300,4 +297,57 @@ void testReplaceNullsFailsOnNullScalar() {
...
+        {  -1,     1,     1,     2,     7},
+        {  -1,     1,     1,     2,     7},
+        {-1.0,   1.0,   1.0,   2.0,   7.0},
+        {  -1,     1,     1,     2,     7}};
+    double[] q = {0.0, 0.25, 0.33, 0.5, 1.0};
+
+    try (ColumnVector cv = ColumnVector.fromBoxedInts(7, 0, 3, 4, 2, 1, -1, 1, 6)) {
+      for (int j = 0 ; j < q.length ; j++) {
+        Scalar result = cv.approxQuantile(q[j]);
+        assertEquals(approxExpected[j], result.getDouble(), 0.000000001);
```
Comment (revans2) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: If we are getting a double out, why are we comparing it to an int?

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 3]
```diff
@@ -300,4 +297,57 @@ void testReplaceNullsFailsOnNullScalar() {
...
+    int[] approxExpected = {-1, 1, 1, 2, 7};
+    double[][] exactExpected = {
+        {-1.0,   1.0,   1.0,   2.0,   7.0},
+        {  -1,     1,     1,     2,     7},
+        {  -1,     1,     1,     2,     7},
+        {-1.0,   1.0,   1.0,   2.0,   7.0},
+        {  -1,     1,     1,     2,     7}};
+    double[] q = {0.0, 0.25, 0.33, 0.5, 1.0};
+
+    try (ColumnVector cv = ColumnVector.fromBoxedInts(7, 0, 3, 4, 2, 1, -1, 1, 6)) {
```
Comment (revans2) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: Can we do this with an even number of entries so there is some kind of a difference in the quantile calculations?

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 4]
```diff
@@ -300,4 +297,57 @@ void testReplaceNullsFailsOnNullScalar() {
...
+  @Test
+  void testIntegerQuantiles() {
+    assumeTrue(Cuda.isEnvCompatibleForTesting());
+    int[] approxExpected = {-1, 1, 1, 2, 7};
+    double[][] exactExpected = {
+        {-1.0,   1.0,   1.0,   2.0,   7.0},
+        {  -1,     1,     1,     2,     7},
+        {  -1,     1,     1,     2,     7},
+        {-1.0,   1.0,   1.0,   2.0,   7.0},
+        {  -1,     1,     1,     2,     7}};
```
Comment (revans2) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: Could we at least label the different methods?  It is hard to tell what each one is expected to be testing?  I get that we want this kind of data-driven, but I think having clarity in the tests might be preferable.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 5]
```diff
@@ -300,4 +298,58 @@ void testReplaceNullsFailsOnNullScalar() {
...
+
+    try (ColumnVector cv = ColumnVector.fromBoxedInts(7, 0, 3, 4, 2, 1, -1, 1, 6, 9)) {
+      // sorted: -1, 0, 1, 1, 2, 3, 4, 6, 7, 9
+      for (int j = 0 ; j < quantiles.length ; j++) {
+        Scalar result = cv.approxQuantile(quantiles[j]);
+        assertEquals(approxExpected[j], result.getInt());
+
+        for (int i = 0 ; i < methods.length ; i++) {
+          result = cv.exactQuantile(methods[i], quantiles[j]);
+          assertEquals(exactExpected[i][j], result.getDouble(), 0.000000001);
```
Comment (abellina) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: small nit, could we define `0.000000001` as a constant somewhere?

**Comment by razajafri** [comment]
@revans2 I have addressed your concerns PTAL

File: CHANGELOG.md [hunk 6]
```diff
@@ -81,6 +81,7 @@
 - PR #1845 Add cudf::drop_duplicates, DataFrame.drop_duplicates
 - PR #1652 Added `Series.where()` feature
 - PR #2074 Java Aggregates, logical ops, and better RMM support
+- PR #2179 Added quantiles
```
Comment (abellina) [inline CHANGELOG.md]: this is listed in 0.8, we should move above to 0.9.

Comment (razajafri) [reply CHANGELOG.md]: good catch

**Pull request merged**